features/net/network-socket/nsapi_types.h

```diff
@@ -148,6 +148,18 @@ typedef enum nsapi_option {
 } nsapi_option_t;
 
 
+/*  C linkage definition
+ *
+ *  Attribute garuntees emitted function has C linkage.
+ *  Available in both C and C++
+ */
+#ifdef __cplusplus
+#define NSAPI_C_LINKAGE extern "C"
```

Comment:
We do in C header files files:

```
#ifdef __cplusplus
extern "C" {
#endif

//declarations of C functions here

#ifdef __cplusplus
}
#endif
```

I don't think this is needed to hide behind a macro. If we would need to have it for each declaration, then we shall consider it, but as a group, this is good as it is.